ouroboros-consensus/test/storage-test/Test/Ouroboros/Storage/LedgerDB/StateMachine.hs

```diff
@@ -0,0 +1,529 @@
+        atomically (forkerCommit forker)
+        forkerClose forker
+      _ -> error "Couldn't restart the chain, failed to apply volatile blocks!"
+  pure (ldb, od)
+
+{-------------------------------------------------------------------------------
+  RunModel
+-------------------------------------------------------------------------------}
+
+-- | The environment for the monad in which we will run the test
+data Environment =
+  Environment
+    (LedgerDB' IO TestBlock)
+    (OnDemandActions' IO TestBlock)
+    (ChainDB IO)
+    (SecurityParam -> IO (TestArguments IO))
+    (IO ())
+
+instance RunModel Model (StateT Environment IO) where
+
+  perform _ (Init secParam) _ = do
+    Environment _ _ chainDb mkArgs _ <- get
+    (ldb, od, clean) <- lift $ do
+      args <- mkArgs secParam
+      (ldb, od) <- openLedgerDB (argFlavorArgs args) chainDb (extLedgerDbConfig secParam) (argHasFS args)
+      pure (ldb, od, cleanup args)
+    put (Environment ldb od chainDb mkArgs (closeBackingStore od >> clean))
+  perform _ (DoTruncate s) lk = do
+    Environment _ od _ _ _ <- get
+    lift $ truncateSnapshot od (lk s)
+  perform _ (DoDelete s) lk = do
+    Environment _ od _ _ _ <- get
+    lift $ destroySnapshot od (lk s)
+  perform _ GetState _ = do
+    Environment ldb _ _ _ _ <- get
+    lift $ atomically $ (,) <$> getImmutableTip ldb <*> getVolatileTip ldb
+  perform _ ForceTakeSnapshot _ = do
+    Environment ldb od _ _ _ <- get
+    lift $ do
+      void $ flushNOW od >> takeSnapshotNOW od Nothing
+      imm <- atomically $ getImmutableTip ldb
+      let pt = pointToWithOriginRealPoint $ castPoint @(ExtLedgerState TestBlock) @TestBlock $ getTip imm
+      pure
+         $ withOrigin Nothing (\p -> Just (DiskSnapshot (unSlotNo $ realPointSlot p) Nothing))
+         $ pt
+  perform _ (ValidateAndCommit n blks) _ = do
+      Environment ldb _ chainDb _ _ <- get
+      lift $ do
+        atomically $ modifyTVar (dbBlocks chainDb) $
+           repeatedly (uncurry Map.insert) (map (\b -> (blockRealPoint b, b)) blks)
+        withRegistry $ \rr -> do
+          vr <- validate ldb rr (const $ pure ()) BlockCache.empty n (map getHeader blks)
+          case vr of
+            ValidateSuccessful forker -> do
+              atomically $ modifyTVar (dbChain chainDb) (reverse (map blockRealPoint blks) ++)
+              atomically (forkerCommit forker)
+              forkerClose forker
+            ValidateExceededRollBack{} -> error "Unexpected Rollback"
+            ValidateLedgerError (AnnLedgerError forker _ _) -> forkerClose forker >> error "Unexpected ledger error"
+  perform state@(Model _ _ secParam) (DropAndRestore n) lk = do
+    Environment _ od chainDb _ clean <- get
+    lift $ do
+      atomically $ modifyTVar (dbChain chainDb) (drop (fromIntegral n))
+      closeBackingStore od
+      clean
+    perform state (Init secParam) lk
+  perform UnInit _ _ = error "Uninitialized model created a command different than Init"
```

Comment:
Consider factoring out each case into its own function, which would make the code easier to read IMO